starfish/imagestack/parser/crop.py

```diff
@@ -0,0 +1,198 @@
+from starfish.imagestack.physical_coordinate_calculator import recalculate_physical_coordinate_range
+from starfish.types import Coordinates, Indices, Number
+
+
+class CropParameters:
+    """Parameters for cropping an ImageStack at load time."""
+    def __init__(
+            self,
+            *,
+            permitted_rounds: Optional[Collection[int]]=None,
+            permitted_chs: Optional[Collection[int]]=None,
+            permitted_zlayers: Optional[Collection[int]]=None,
+            x_slice: Optional[Union[int, slice]]=None,
+            y_slice: Optional[Union[int, slice]]=None,
+    ):
+        """
+        Parameters
+        ----------
+        permitted_rounds : Optional[Collection[int]]
+            The rounds in the original dataset to load into the ImageStack.  If this is not set,
+            then all rounds are loaded into the ImageStack.
+        permitted_chs : Optional[Collection[int]]
+            The channels in the original dataset to load into the ImageStack.  If this is not set,
+            then all channels are loaded into the ImageStack.
+        permitted_zlayers : Optional[Collection[int]]
+            The z-layers in the original dataset to load into the ImageStack.  If this is not set,
+            then all z-layers are loaded into the ImageStack.
+        x_slice : Optional[Union[int, slice]]
+            The x-range in the x-y tile that is loaded into the ImageStack.  If this is not set,
+            then the entire x-y tile is loaded into the ImageStack.
+        y_slice : Optional[Union[int, slice]]
+            The y-range in the x-y tile that is loaded into the ImageStack.  If this is not set,
+            then the entire x-y tile is loaded into the ImageStack.
+        """
+        self._permitted_rounds = set(permitted_rounds) if permitted_rounds else None
+        self._permitted_chs = set(permitted_chs) if permitted_chs else None
+        self._permitted_zlayers = set(permitted_zlayers) if permitted_zlayers else None
+        self._x_slice = x_slice
+        self._y_slice = y_slice
+
+    def filter_tilekeys(self, tilekeys: Collection[TileKey]) -> Collection[TileKey]:
+        """
+        Filters tilekeys for those that should be included in the resulting ImageStack.
+        """
+        results: MutableSequence[TileKey] = list()
+        for tilekey in tilekeys:
+            if self._permitted_rounds is not None and tilekey.round not in self._permitted_rounds:
+                continue
+            if self._permitted_chs is not None and tilekey.ch not in self._permitted_chs:
+                continue
+            if self._permitted_zlayers is not None and tilekey.z not in self._permitted_zlayers:
+                continue
+
+            results.append(tilekey)
+
+        return results
+
+    @staticmethod
+    def _crop_axis(size: int, crop: Optional[Union[int, slice]]) -> Tuple[int, int]:
+        """
+        Given the size of along an axis, and an optional cropping, return the start index
```

Comment:
given an edge of size N, and a set of cropping parameters, return the start and and end indices.  This is to support things like where the cropping parameters are outside the bounds of the size, or negative cropping parameters like [0, -2] (crop off the last pixel, regardless of size)

Reply:
Got it. This makes sense. 

> [0, -2] (crop off the last pixel, regardless of size)

In case this wasn't a mistake with the example, this is not how python normally works. `list[0:-1]` would crop off the last item, regardless of size. We should be consistent: 

```python
In [1]: x = list('abc')                                                                                                           

In [2]: x[0:-1]                                                                                                                   
Out[2]: ['a', 'b']
```

Reply:
No, that was a mistake.  [0, -2] will crop off _two_ pixels, as you said.